### ISSUE TITLE
Only type-level `safeLink`?

### ISSUE THREAD

**Issue by eyeinsky** [haskell-servant/servant#1449]
There is probably a reason for this, but why isn't it possible to provide `safeLink` with two proxies upfront to make a link-maker like this:
```
apiLink :: forall endpoint. (IsElem endpoint API, HasLink endpoint) => MkLink endpoint Link
apiLink = safeLink (Proxy :: Proxy API) (Proxy :: Proxy endpoint)
```

So it could be used like this: `apiLink @SomeRoute`?

**Comment by eyeinsky** [comment]
Thanks @considerate!

@alpmestan What I was getting at was that couldn't the `safeLink` be pre-applied by `Proxy` values and one could just fill the type arguments directly at use site. Now with `AllowAmbiguousTypes` I was able to come up with the following which does appear to work:
```
aLink
  :: forall route server.
  ( MkLink route Link ~ Link
  , IsElem route server
  , HasLink route)
  => Link
aLink = safeLink (Proxy @server) (Proxy @route)
```

It could be used like so: `let link = aLink @RouteInAPI @FullAPI in ...`

Not an expert in type level programming but brief research seems to indicate that `AllowAmbiguousTypes` can be useful together with `TypeApplications` (and `ScopedTypeVariables`).

**Comment by considerate** [comment]
> Not an expert in type level programming but brief research seems to indicate that AllowAmbiguousTypes can be useful together with TypeApplications (and ScopedTypeVariables).

Yes I agree that it can be useful. The issue with `AllowAmbiguousTypes` is that the error messages might end up in unexpected places since you propagate the ambiguity away from the call site so in larger programs it can be a bit more difficult to understand why something fails to typecheck. However, if it's clear that `aLink` should always be used with both types visibly applied with `TypeApplications` then I think it can make sense to add this language extension.

Here's [a discussion about the danger of using `AllowAmbiguousTypes`](https://stackoverflow.com/questions/49684655/how-dangerous-is-allowambiguoustypes-when-used-with-typeapplications) for context.
